From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Wed, 27 Jul 2022 01:51:15 +0900
Subject: [PATCH] Purpur PR - Add ability to place and remove block debug
 markers for players

Original by YouHaveTrouble <youhavetrouble@youhavetrouble.me> (PurpurMC/Purpur#1072)

Copyright (c) 2022 PurpurMC LLC

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 8bf04a0ad920fa580568713b243c20c448435908..baa86a63e4654242c4dfc12123540feea41f05b5 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2832,5 +2832,53 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @return True if fall damage is enabled when {@link #getAllowFlight()} is true
      */
     public boolean hasFlyingFallDamage();
+
+    // Prismarine start - Purpur PR
+    /**
+     * Creates debug block highlight on specified block location. Clients may be inconsistant in displaying it.
+     * @param location Location to highlight
+     * @param text Text to show above the highlight
+     * @param time Time after which highlight will expire in milliseconds
+     * @param color RGB int. Will be ignored on some versions of vanilla client
+     */
+    public void sendBlockHighlight(@NotNull Location location, @NotNull String text, int time, int color);
+
+    /**
+     * Creates debug block highlight on specified block location. Clients may be inconsistant in displaying it.
+     * @param location Location to highlight
+     * @param time Time after which highlight will expire in milliseconds
+     * @param color RGB int. Will be ignored on some versions of vanilla client
+     */
+    public void sendBlockHighlight(@NotNull Location location, int time, int color);
+
+    /**
+     * Creates debug block highlight on specified block location. Clients may be inconsistant in displaying it.
+     * @param location Location to highlight
+     * @param time Time after which highlight will expire in milliseconds
+     * @param color Color of the highlight. Will be ignored on some versions of vanilla client
+     * @param transparency Transparency of the highlight. Values not in range of 0-255 will throw IllegalArgumentException.
+     */
+    public void sendBlockHighlight(@NotNull Location location, int time, @NotNull org.bukkit.Color color, int transparency);
+
+    /**
+     * Creates debug block highlight on specified block location. Clients may be inconsistant in displaying it.
+     * @param location Location to highlight
+     * @param text Text to show above the highlight
+     * @param time Time after which highlight will expire in milliseconds
+     */
+    public void sendBlockHighlight(@NotNull Location location, @NotNull String text, int time);
+
+    /**
+     * Creates debug block highlight on specified block location. Clients may be inconsistant in displaying it.
+     * @param location Location to highlight
+     * @param time Time after which highlight will expire in milliseconds
+     */
+    public void sendBlockHighlight(@NotNull Location location, int time);
+
+    /**
+     * Clears all debug block highlights created by Player#sendBlockHighlight.
+     */
+    public void clearBlockHighlights();
+    // Prismarine end
     // Purpur end
 }
